File: aat-lib/src/main/java/ch/bailu/aat_lib/preferences/SolidString.java
package ch.bailu.aat_lib.preferences;

import java.util.ArrayList;

import ch.bailu.aat_lib.exception.ValidationException;

public class SolidString extends AbsSolidType {
    private final String key;
    private final StorageInterface storage;


    public SolidString(StorageInterface storage, String key) {
        this.storage = storage;
        this.key = key;
    }

    @Override
    public String getValueAsString() {
        return getStorage().readString(getKey());
    }

    @Override
    public void setValueFromString(String s) throws ValidationException {
        setValue(s);
    }

    public void setValue(String v) {
        getStorage().writeString(key, v);
    }


    @Override
    public String getKey() {
        return key;
    }

    @Override
    public StorageInterface getStorage() {
        return storage;
    }


    public ArrayList<String> buildSelection(ArrayList<String> strings) {
        return strings;
    }

    public String getValueAsStringNonDef() {
        String s = getValueAsString();
        if (s == null || storage.isDefaultString(s)) return "";
        return s;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/preferences/SolidString.kt
package ch.bailu.aat_lib.preferences

import ch.bailu.aat_lib.exception.ValidationException

open class SolidString(private val storage: StorageInterface, private val key: String) : AbsSolidType() {

    override fun getValueAsString(): String {
        return getStorage().readString(getKey())
    }

    @Throws(ValidationException::class)
    override fun setValueFromString(string: String) {
        setValue(string)
    }

    open fun setValue(v: String) {
        getStorage().writeString(key, v)
    }

    override fun getKey(): String {
        return key
    }

    override fun getStorage(): StorageInterface {
        return storage
    }

    /**
     * List of possible values to select a value from
     * This is used by selection lists in the settings dialog
     */
    open fun buildSelection(list: ArrayList<String> = ArrayList(5)): ArrayList<String> {
        return list
    }

    fun getValueAsStringNonDef(): String {
        val s = getValueAsString()
        return if (storage.isDefaultString(s)) "" else s
    }

    fun getDefaultValueFromSelection(fallBack: String = getStorage().getDefaultString()): String {
        val list = buildSelection()
        list.add(fallBack)
        return list[0]
    }

    fun setDefaultValue(): String {
        val r = getDefaultValueFromSelection()
        setValue(r)
        return r
    }
}
